doc/developer/design/20220516_transactional_consistency.md

```diff
@@ -0,0 +1,446 @@
+# Materialize Transactional Consistency
+
+## Summary
+
+A description of Materialize’s consistency guarantees with respect to the operations that occur inside of Materialize.
+Operations include:
+
+- `SELECT`, `INSERT`, `UPDATE`, and `DELETE` statements (but not `TAIL`)
+
+This design document is in service of [Epic #11631](https://github.com/MaterializeInc/materialize/issues/11631)
+and [Correctness Property #1](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/platform/ux.md#correctness)
+. Specifically for the first bullet point from the correctness property, “`SELECT`, `INSERT`, `UPDATE`, and `DELETE`
+statements (but not `TAIL`)". A follow-up design document will be created for the second bullet point, "Materialize
+initiated acknowledgements of upstream sources (e.g., the commit of an offset by a Kafka source, the acknowledge of an
+LSN by a PostgresSQL source, etcetera)".
+
+This document does not take any opinion on whether the changes proposed should be the default behavior or an opt-in
+behavior via some configuration.
+
+## Goals
+
+The goal of this design document is to describe the transactional consistency guarantees that are provided by
+Materialize and describe what changes are needed to provide higher levels of transaction consistency. This document will
+focus on Read Uncommitted, Read Committed, Repeatable Read, Serializable, and Strict Serializable consistency levels.
+
+## Non-Goals
+
+This design document does not focus on levels of consistency not mentioned in the [Goals](#Goals) sections unless it is
+in service of achieving one of the mentioned consistency levels.
+
+Sinks are left out of this document and not considered for consistency.
+
+`TAIL` is left out of this document and not considered for consistency.
+
+`AS OF` allows clients to specify the timestamp of a query. This allows clients to circumvent our consistency
+guarantees, so they’re left out of this document.
+
+This document does not include Materialize initiated acknowledgements of upstream sources (e.g., the commit of an offset
+by a Kafka source, the acknowledgment of an LSN by a PostgresSQL source, etcetera). For the purposes of this document we
+will consider all of these events to have happened at the timestamp assigned to them by timely workers. This is not
+accurate for multiple reasons, some of which include:
+
+- Clock Skew: Each timely worker has their own clock that is used to timestamp events, that may be out of sync with
+  other clocks.
+- Event Time vs Process Time: Many events are timestamped with the time they arrive at Materialize not when they
+  actually happen. Network and other delays may make this timestamp inaccurate.
+- Reading from these sources almost always happen at a timestamp lower than what was just written to these source.
+- Etc.
```

Comment:
In my interpretation receiving data was directly related to acknowledging data because Materialize would be acknowledging data directly after receiving data and making it durable. Perhaps that isn't an accurate interpretation though. 

I think the scenario that I was trying to point out is: 
1. MZ receives data from some source, ingests the data, gives it a timestamp of `ts`, and sends an acknowledgement.
2. Some other MZ nodes receives data from some other source, ingests that data, also gives it a timestamp of `ts`, and sends an acknowledgement.

MZ may have actually received those two pieces of data at different times due to clock skew, and those the two pieces of data may have originally been generated at different times due to event time vs process time.